### DISCUSSION TITLE
How to stop a running exposure?

### DISCUSSION THREAD

**Discussion by scriptorron** [raspberrypi/picamera2#589] [Q&A] [answered]
First of all many thanks for providing this great software!

I use picamera2 for astro-photography. For these applications I am only interested in the raw Bayer data. It is typical to have extreme long exposure times (minutes!). 

In some situations it is needed to reconfigure or restart the camera. But when the camera was already started with a long exposure time the `stop()` will block until end of the running exposure. A test bench  for Raspi HQ camera is here:

```python
import time
from picamera2 import Picamera2
from libcamera import controls

print('Opening and configuring camera...')

picam2 = Picamera2()
config = picam2.create_still_configuration(
    queue=False,
    buffer_count=1
)
config["main"]["size"] = (240, 190)
config["raw"] = {"size": (4056, 3040), "format": 'SRGGB12'}

picam2.align_configuration(config)
picam2.configure(config)
picam2.set_controls(
    {
        # controls for main frame: disable all regulations
        "AeEnable": False,  # AEC/AGC algorithm
        # disable noise reduction in main frame because it eats stars
        "NoiseReductionMode": controls.draft.NoiseReductionModeEnum.Off,
        # disable automatic white balance algorithm and set colour gains manually
        "AwbEnable": False,
        "ColourGains": (2.0, 2.0),  # to compensate the 2 G pixel in Bayer pattern
        # low level controls
        "AnalogueGain": 20,
        "ExposureTime": int(60 * 1e6),  # 1 minute
        # exposure time in us; needs to be integer!
    }
)

StartTime = time.time()
print(f'Starting camera at T=0s')
picam2.start()

time.sleep(1.0)

print(f'Found that I need to stop camera at T={time.time() - StartTime:.2f}s')
picam2.stop()

print(f'Camera stopped at T={time.time() - StartTime:.2f}s')
```
It generates this output:
```
Opening and configuring camera...
[2:06:51.905291314] [1797]  INFO Camera camera_manager.cpp:299 libcamera v0.0.2+55-5df5b72c
[2:06:52.063278958] [1798]  INFO RPI raspberrypi.cpp:1425 Registered camera /base/soc/i2c0mux/i2c@1/imx477@1a to Unicam device /dev/media3 and ISP device /dev/media0
[2:06:52.109409561] [1797]  INFO Camera camera.cpp:1026 configuring streams: (0) 224x190-BGR888 (1) 4056x3040-SBGGR12
[2:06:52.112531534] [1798]  INFO RPI raspberrypi.cpp:805 Sensor: /base/soc/i2c0mux/i2c@1/imx477@1a - Selected sensor format: 4056x3040-SBGGR12_1X12 - Selected unicam format: 4056x3040-BG12
Starting camera at T=0s
Found that I need to stop camera at T=1.11s
Camera stopped at T=60.80s
```

You can see that I call `stop()` at T=1.11s. But that takes until T=60.80s to execute. In this example I used only 1 minute exposure time, for astro-photography it is not unusual to have 3 or 4 minutes (as long as the camera supports).

In such situations the running exposure will anyway thrown away. Waiting until the end of the exposure is a wast of time. 

Is there a way to stop the camera faster?
Terminating the Python program is not an option because I need to start the camera later again with a new configuration.

Can I use `picam2.stop_()` instead of `picam2.stop()`? This stops immediately, but its docstring says "Only call this function directly from within the camera event loop, such as in a Qt application.".

**Comment by davidplowman** [comment, marked as answer]
Hi, thanks for the question.

When the camera is running, we normally delegate camera-related operations to the event loop which is why the "official" `picam2.stop()` has to wait for the event loop to run again (when a frame turns up). This is probably something we could review for a future release.

In the meantime, I would expect that what you're doing will probably be OK so long as there aren't any other camera operations in progress and you wait a few moments before doing anything else. Though this seems a slightly grey area!

**Reply by scriptorron** [reply]
Many thanks for the answer. I will use `picam2.stop_()`. It works and I do not see any side effects.
